Failed to reap snapshot ...\snapshots\1-1-1510819823632: remove ...\snapshots\1-1-1510819823632\state.bin: The process cannot access  the file because it is being used by another process on Windows

This problem only occurs on Windows. 

I am using `RecoverCluster()` to "restart" a cluster when the number of nodes decreases to 1 and it is not able to elect itself as a leader.

I have a `time.Ticker` ticking at every 5 seconds to check if there is a leader and whether the last time we contacted a leader is past a certain threshold:

```go
if r.Leader() == "" && time.Since(r.LastContact()) > 10*time.Second {
	fmt.Println("Time to restart!!")

	f := r.Shutdown()

	if f.Error() != nil {
		log.Fatalf("Error shutting down raft: %s", f.Error())
	}

	trans, err := raft.NewTCPTransport(raftAddr, nil, 3, 10*time.Second, os.Stdout)

	fmt.Println("Recovering")
	err = raft.RecoverCluster(c, &fsm{}, raftDB, raftDB, snapshotStore, trans, raft.Configuration{
		Servers: []raft.Server{
			{
				Suffrage: raft.Voter,
				ID:       raft.ServerID(raftAddr),
				Address:  raft.ServerAddress(raftAddr),
			},
		},
	})

	if err != nil {
		log.Fatalf("Error recovering cluster: %s", err)
	}
	fmt.Println("starting new raft")
	r, err = raft.NewRaft(c, &fsm{}, raftDB, raftDB, snapshotStore, trans)

	if err != nil {
		log.Fatal(err)
	}
}
```
The problem is that the restart works correctly, but snapshotting fails at a later point if a node leaves the cluster again.

Here's what happens:
1. Start first node (self-elected as master). raft-sample -serfPort=1111
2. Start second node. Serf discovers it and the first node (leader) adds it to raft pool. raft-sample -serfPort=2222 -members=172.28.128.1:1111
3. Start third node. Serf discovers it and the first node (leader) adds it to raft pool. raft-sample -serfPort=3333 -members=172.28.128.1:1111
4. Kill first node and second node is elected as leader.
5. Kill second node and wait for the there to be no leader and the last leader was contacted more than 10 seconds ago. 
6. Shutdown raft, create new transport, `RecoverCluster()` and create new raft in the third node.
7. Third node becomes the leader.
8. Start up fourth node on a blank slate (no previous raft history).
9. Either the fourth node or third node is elected leader.
10. Kill either the fourth node or the third node.
11. The remaining node will see no leader and once the threshold is met, attempt recovery and restart raft. During `RecoverCluster()`, a snapshot error occurs:

````
017/11/16 10:14:53 [INFO] snapshot: Creating new snapshot at C:\work\rafttest\c8b0beac0a5ee0df18505474cad51749\snapshots\14-9-1510787693498.tmp
2017/11/16 10:14:53 [INFO] snapshot: reaping snapshot C:\work\rafttest\c8b0beac0a5ee0df18505474cad51749\snapshots\5-7-1510787663588
2017/11/16 10:14:53 [ERR] snapshot: Failed to reap snapshot C:\work\rafttest\c8b0beac0a5ee0df18505474cad51749\snapshots\5-7-1510787663588: remove C:\work\rafttest\c8b0beac0a5ee0df18505474cad51749\snapshots\5-7-1510787663588\state.bin: The process cannot access the file because it is being used by another process.
2017/11/16 10:14:53 Error recovering cluster: failed to finalize snapshot: remove C:\work\rafttest\c8b0beac0a5ee0df18505474cad51749\snapshots\5-7-1510787663588\state.bin: The process cannot access the file because it is being used by another process.
````

I have simplified the test case further to emulate the snapshot operations in `RecoverCluster()`:

```go
package main

import (
	"log"
	"os"

	"github.com/hashicorp/raft"
)

func main() {

	workDir, err := os.Getwd()

	if err != nil {
		log.Fatal(err)
	}

	_, trans := raft.NewInmemTransport(raft.ServerAddress(""))

	snaps, err := raft.NewFileSnapshotStore(workDir, 1, os.Stdout)

	if err != nil {
		log.Fatalf("Error creating file snapshot store: %s", err)
	}

	sink, err := snaps.Create(1, 1, 1, raft.Configuration{}, 1, trans)

	_, err = sink.Write([]byte("this is a test"))

	if err != nil {
		log.Fatalf("Error writing to sink: %s", err)
	}

	err = sink.Close()

	if err != nil {
		log.Fatalf("Error closing sink: %s", err)
	}

	err = snaps.ReapSnapshots()

	if err != nil {
		log.Fatalf("Error reaping snapshots: %s", err)
	}

	snapshots, err := snaps.List()

	if err != nil {
		log.Fatalf("Error listing snapshots: %s", err)
	}

	for _, snapshot := range snapshots {
		_, source, err := snaps.Open(snapshot.ID)

		if err != nil {
			log.Fatalf("Error opening snapshot: %s", err)
		}

		defer source.Close()
	}

	sink, err = snaps.Create(1, 1, 1, raft.Configuration{}, 1, trans)

	_, err = sink.Write([]byte("this is a test"))

	if err != nil {
		log.Fatalf("Error writing to sink: %s", err)
	}

	err = sink.Close()

	if err != nil {
		log.Fatalf("Error closing sink: %s", err)
	}

	err = snaps.ReapSnapshots()

	if err != nil {
		log.Fatalf("Error reaping snapshots: %s", err)
	}

}

```

Running the test case produces the error:

```
2017/11/16 19:10:23 [INFO] snapshot: Creating new snapshot at C:\Work\rafttest\tt2\snapshots\1-1-1510819823632.tmp
2017/11/16 19:10:23 [INFO] snapshot: Creating new snapshot at C:\Work\rafttest\tt2\snapshots\1-1-1510819823785.tmp
2017/11/16 19:10:23 [INFO] snapshot: reaping snapshot C:\Work\rafttest\tt2\snapshots\1-1-1510819823632
2017/11/16 19:10:23 [ERR] snapshot: Failed to reap snapshot C:\Work\rafttest\tt2\snapshots\1-1-1510819823632: remove C:\Work\rafttest\tt2\snapshots\1-1-1510819823632\state.bin: The process
cannot access the file because it is being used by another process.
panic: [ERR] snapshot: Failed to reap snapshot C:\Work\rafttest\tt2\snapshots\1-1-1510819823632: remove C:\Work\rafttest\tt2\snapshots\1-1-1510819823632\state.bin: The process cannot access
 the file because it is being used by another process.

goroutine 1 [running]:
rafttest/vendor/github.com/hashicorp/raft.(*FileSnapshotStore).ReapSnapshots(0xc0420022e0, 0x58, 0xc04200e6c0)
        C:/Work/rafttest/vendor/github.com/hashicorp/raft/file_snapshot.go:361 +0x542
rafttest/vendor/github.com/hashicorp/raft.(*FileSnapshotSink).Close(0xc04216e0e0, 0xc042008770, 0xe)
        C:/Work/rafttest/vendor/github.com/hashicorp/raft/file_snapshot.go:426 +0x4af
main.main()
        C:/Work/rafttest/tt2/main.go:70 +0x747
exit status 2
```

--- a/api.go
+++ b/api.go
@@ -294,9 +294,11 @@ func RecoverCluster(conf *Config, fsm FSM, logs LogStore, stable StableStore,
 			// couldn't open any snapshots.
 			continue
 		}
-		defer source.Close()
 
-		if err := fsm.Restore(source); err != nil {
+		err = fsm.Restore(source)
+		// Close the source after the restore has completed
+		source.Close()
+		if err != nil {
 			// Same here, skip and try the next one.
 			continue
 		}
@@ -545,9 +547,11 @@ func (r *Raft) restoreSnapshot() error {
 			r.logger.Error(fmt.Sprintf("Failed to open snapshot %v: %v", snapshot.ID, err))
 			continue
 		}
-		defer source.Close()
 
-		if err := r.fsm.Restore(source); err != nil {
+		err = r.fsm.Restore(source)
+		// Close the source after the restore has completed
+		source.Close()
+		if err != nil {
 			r.logger.Error(fmt.Sprintf("Failed to restore snapshot %v: %v", snapshot.ID, err))
 			continue
 		}

